### ISSUE TITLE
Use cluster by keys in merge for faster incremental updates

### ISSUE THREAD

**Issue by bastienboutonnet** [fishtown-analytics/dbt#2139]
### Describe the feature
#1689 brought the ability for users to leverage Snowflake's clustering in dbt.
We recently realised that maybe on clustered tables we could make merge operations faster by using the cluster keys provided on the model. After testing it via a custom materialisation in our macros we really saw a huge improvement!

### Additional context
Snowflake specific

### Who will this benefit?
Anyone on snowflake with large tables who leverages clustering in dbt.

I'd be more than happy to contribute this functionality back to dbt as it's fairly simple. 
It would basically consist in adding a test for whether `cluster_by_keys` are provided and adding them kind of like the following snipped.

```
        {% if cluster_by_keys is not none %}
            {% for key in cluster_by_keys %}
                and DBT_INTERNAL_SOURCE.{{key}} = DBT_INTERNAL_DEST.{{key}}
            {% endfor %}
        {% endif %}
```

If this is something you guys would be insterested in I can put up a PR fairly quickly. Would love, first to hear your thoughts, and discuss potential caveats.

**Comment by pedromachados** [comment]
I noticed something similar but haven't tried your approach. I started to use surrogate keys that are simply the concatenation of fields (no md5 applied to the concatenated string). The first columns in those keys are the cluster keys. This improves the merge performance (partition pruning). It also achieves better compression so reads are probably faster.

Could there be a problem with your approach if the cluster keys are not part of the surrogate key? Trying to wrap my mind around this.

**Comment by bastienboutonnet** [comment]
@pedromachados If I understand well, in your use case you build a surrogate and use that surrogate to cluster? I am always in favour of surrogates for a variety of reasons. So I'm not going to argue against it, however, depending what you create that surrogate for you might end up putting into it stuff with quite a high cardinality and I believe using this to cluster may not give you a lot of benefits since you'd end up having very unique partitions. Probably still a benefit but I would argue less large than using stuff with low cardinality. 

So let's consider a dummy use case:
If you have a series of events that come in at pretty large volume. Each of these events are/should be unique and are identified with a `uuid`. Using the uuids to cluster your table would result in creating partitions that are as plentiful as you have events. If you were to surrogate say `loaded_on` (date) with `event_uuid` I don't know nor think there would be a huge benefit here since your hash is still going to be `n = event_uuid`.

In that case, while your surrogate might help you validating some assumptions etc. in your model, it wouldn't help your clustering.

The use case we tested this on is one where:
- table is clustered on `loaded_on`. Which is a date and thus much lower cardinality than the number of records we load.
- primary key is `event_uuid`
- if for some reason `event_uuid` was not unique let's say event uuids are shared between users you may need to implement a surrogate to capture that uniqueness with `event_uuid, user_uuid`.

Now, you want to perform a merge. The only thing you really need is to identify your unique record. This could be the `event_uuid` or your `event_user_key` surrogate. You could benefit from a speed up, if you have clustered your table by date and feed that to the list of keys the merge uses so that the merge can directly target the partition where that unique record is (avoiding a full table scan), looking for that unique record in that reduced space and performing the data update.

But to be completely honest, I don't know how Snowflake really deals with clustering and whether it likes low cardinality better or whether it's intelligent enough not to care. Would be more than happy if anyone had some info on this actually.

So regarding this question:
> Could there be a problem with your approach if the cluster keys are not part of the surrogate key? Trying to wrap my mind around this.
I think it depends how your table is clustered. If your table is clustered on say `loaded_on` (only) and you merge using `surrogate_key` as your `primary_key` you wouldn't get a benefit from partition pruning since your table isn't clustered like that. If your table is clustered on your `surrogate_key` and you use that key as your `primary_key` chances are you should see a benefit (provided that the cardinality aspect mentioned above doesn't hurt the effectiveness of the clustering)

Ultimately, in my opinion and design habits, I consider clustering/partitioning keys and primary keys as very different things serving very different purposes. - - Primary key being a way to identify records uniquely, tests assumptions, validate integrity etc
- Clustering keys, purely a database optimisation thing.
Now in some cases they may overlap, and in some they won't which is fine as they serve a different purpose.

Do you by that?

**Comment by pedromachados** [comment]
Hi @bastienboutonnet I agree with your points. 

In our case, we are clustering on `date` and `country` (in that order) but our `surrogate_key` is roughly defined as `date || '-' || country || '-' || market_id`. So the start of our `surrogate_key` is correlated with the cluster keys. We did not add the `market_id` to the cluster key because Snowflake does not recommend high-cardinality columns in the cluster key in order to avoid the problem you mention (too many partitions).

We also found that we get better performance with this approach than when we hash the key as in `md5(date || '-' || country || '-' || market_id)`. I suppose (but don't know for sure) this has to do with compression (consecutive values have very different hash keys and don't compress well) as well as the fact that if you hash the key Snowflake may not be able to figure out that it can prune partitions in the case of an incremental model run. In this case, I think that adding `date`, `country`, and `market_id` to the `merge` condition as you suggest would help even if you have a hashed key because Snowflake has enough information to understand where those new records go.

My key question about your suggestion is whether it changes the logic of the merge if the cluster keys are not part of the surrogate key. For example, suppose you have a table where you intend to have unique records by `uuid` but you get the same `uuid` record in two different `loaded_on` dates. If you add `loaded_on` to the merge condition, would you not end up with two records with the same `uuid`?

**Comment by bastienboutonnet** [comment]
2 first paragraphs make a lot of sense, and interesting find around the `md5` shit we actually use md5s a lot in our surrogates. 

For your last point. I feel, maybe I'm dumb and not understanding your issue. So let me rephrase. If you were to load records where `uuid` should be unique but you can have the same uuid on two `loaded_on` dates then I would say it seems there is an issue with the assumption that uuid is unique. It seems, that it that can be the case that you would **want* to load it twice (because it's not unique) or treat before inserting so you enforce uniqueness because this double loaded on is a fluke. I think if you were not to add the loaded_on in the merge you'd get a DML error from Snowflake telling you it found a duplicate record with that primary key. Worth testing I guess...

**Comment by drewbanin** [comment]
Hey y'all - really cool discussion!

I think @pedromachados has a good point -- what happens when the `unique_key` is not generated solely from the list of clustering keys? Eg:

| id | color | name |
| -  | ----        | ---    |
| 1 | yellow | Drew |
| 2 | green | Pedro |
| 3 | purple | Bastien |

Given:
```sql
{{ config(
  materialized='incremental',
  unique_key='id',
  cluster_by_keys =['color']
) }}

select * from people
}}
```

In this example, imagine that our source data changes and the resulting model select statement returns:

| id | color | name |
| -  | ----        | ---    |
| 1 | blue | Drew |


The existing merge will look like:
```sql
merge into dest
using source
on dest.id = source.id
when matched then update...
when not matched then insert...
```

and it will serve to update the record with `id=1` to have `color=blue`.

If we change this `merge` logic to use the clustering keys, it would instead look like:
```sql
merge into dest
using source
on dest.id = source.id and dest.color = source.color
when matched then update...
when not matched then insert...
```

Since the `color` field has changed for the record with `id=1,` I think we'd see:
| id | color | name |
| -  | ----        | ---    |
| 1 | yellow | Drew |
| 2 | green | Pedro |
| 3 | purple | Bastien |
| 1 | blue | Drew |

So, I think that we might not want to _automatically_ do this, but I do think it's a pretty good optimization technique if we can figure out a good way to implement it!

**Comment by bastienboutonnet** [comment]
Hmmm!!! I see now I see! In that case it does indeed have a pretty nefarious consequence... It might be the case that we'd have to think about a certain flag that the user could turn on conciously when it makes sense to use the cluster keys (i.e that in combination with the primary keys they will be ensuring uniqueness and therefore somewhat immutable) Vs a case where the field(s) used in clustering may be allowed to mutate. Does that make sense @drewbanin and @pedromachados ?

**Comment by pedromachados** [comment]
What if we added another list where you put `additional_merge_keys` (or a better name)?

```
{{
    config(
        materialized="incremental",
        unique_key="surrogate_key",
        additional_merge_keys=["event_date", "country"],
        cluster_by=["event_date", "country", "platform"],
    )
}}
```

**Comment by drewbanin** [comment]
@pedromachados I like that idea! We have just implemented something similar in BigQuery (#1971) and we're using the term `predicates`. I think this would make sense on any database that uses a `merge` statement.

My vote is for a name that isn't `merge` specific, as I think we can use this same config in incremental strategies that use `delete + insert` statements. Maybe `predicate_columns`? Y'all think that's too vague?

**Comment by pedromachados** [comment]
@drewbanin Just to make sure I understand, you are proposing the name `predicate_columns` to be consistent with the `predicates` argument, but you are not suggesting we use `predicates` for this. Correct?

If we were to use `predicates` the way I understand it the user would have to use the internal DBT aliases `DBT_INTERNAL_DEST` and `DBT_INTERNAL_SOURCE` in the predicate expressions which doesn't seem right.

So if we are adding another argument `predicate_columns`, would this be a list of columns that dbt would add to the merge condition? For example:

`predicate_columns=["predicate_column1"]` would add `... and DBT_INTERNAL_DEST.predicate_column1 = DBT_INTERNAL_SOURCE.predicate_column1` to the merge condition. Correct?

I don't think `predicate_columns` is self-explanatory but we could deal with it in the documentation. Could you comment on how this would apply to the `delete + insert` strategy? Is there a way to use this in the `delete` statement to speed it up or would it be ignored?

**Comment by bastienboutonnet** [comment]
@drewbanin , @pedromachados I think I like the approach. 
It's less covert about what might happen and therefore. I think we'd probably still want to warn people who might have not realised duplicates were being (albeit bruteforce dealt with) with the current implementation that they might come up, why and how to deal with it.

With regards to naming, I find predicates perhaps a little too jargony (although that might be due to a personal bias, I lived for 2+ years of writing SQL without having ever heard anyone refer to these groups of keywords... go figure). To me `merge` is ok, also because we refer to the other option as `merge_strategy` so it would fit but it isn't entirely correct.

@pedromachados with regards to the implementation, I would anticipate that we'd do the following yes: ` DBT_INTERNAL_DEST.predicate_column1 = DBT_INTERNAL_SOURCE.predicate_column1`
I still think the delete would be able to leverage the partitions since the delete approach has a way to could add a series of `and`s judging by the implementation here for example https://github.com/fishtown-analytics/dbt/blob/dev/barbara-gittings/core/dbt/include/global_project/macros/materializations/common/merge.sql#L66-L69

So I guess if we were to agree on naming and all doubts are cleared I'd be happy to get cracking.

**Comment by drewbanin** [comment]
@jtcohen6 check this thread out - curious to hear what you've think w/r/t to some of the BQ merge statement tuning you've done recently

**Comment by louis-vines** [comment]
I've just come across this discussion in 2021 and am very interested. What's the status of this? I'd happliy try and put a PR in myself but am quite new to DBT.

**Comment by bastienboutonnet** [comment]
I'm also still happy to help, I just had completely forgotten about this discussion and it looks like we never reached a conclusion on this.

**Comment by jtcohen6** [comment]
Lately, I'm interested in a flexible (though less opinionated) solution in the form of an `incremental_predicates` config, as proposed in https://github.com/fishtown-analytics/dbt/issues/3293.

In such an implementation, dbt wouldn't be responsible for as much behind-the-scenes magic. It would be up to the user to specify the filtering/matching logic using their cluster key, and pass along as one or more `incremental_predicates`.
